src/creation.cpp

```diff
@@ -0,0 +1,436 @@
+    return S2LatLng::FromDegrees(lat, lng).ToPoint();
+}
+
+S2Point make_s2point(const std::pair<double, double> &point) {
+    return S2LatLng::FromDegrees(point.second, point.first).ToPoint();
+}
+
+S2Point make_s2point(const Geography *point_ptr) {
+    check_geog_type(*point_ptr, GeographyType::Point);
+    auto s2geog_obj = static_cast<const s2geog::PointGeography &>(point_ptr->geog());
+
+    if (s2geog_obj.Points().empty()) {
+        throw EmptyGeographyException("cannot create s2geometry point from empty POINT Geography");
+        // TODO: what do we want here? NaN or S2Point default constructor? It depends?
+        // return S2Point::NaN();
+    }
+
+    return s2geog_obj.Points()[0];
+}
+
+template <class V>
+std::vector<S2Point> make_s2points(const std::vector<V> &points) {
+    std::vector<S2Point> s2points(points.size());
+
+    auto func = [](const V &vertex) {
+        return make_s2point(vertex);
+    };
+
+    std::transform(points.begin(), points.end(), s2points.begin(), func);
+
+    return std::move(s2points);
+}
+
+// create a S2Loop from coordinates or Point objects.
+//
+// Normalization (to CCW order for identifying the loop interior) and validation
+// are both enabled by default.
+//
+// Additional normalization is made here:
+// - if the input loop is already closed, remove one of the end nodes
+//
+// TODO: add option to skip normalization.
+//
+template <class V>
+std::unique_ptr<S2Loop> make_s2loop(const std::vector<V> &vertices, bool check = true) {
+    auto s2points = make_s2points(vertices);
+
+    if (s2points.front() == s2points.back()) {
+        s2points.pop_back();
+    }
+
+    auto loop_ptr = std::make_unique<S2Loop>();
+
+    loop_ptr->set_s2debug_override(S2Debug::DISABLE);
+    loop_ptr->Init(s2points);
+    if (check && !loop_ptr->IsValid()) {
+        std::stringstream err;
+        S2Error s2err;
+        err << "ring is not valid: ";
+        loop_ptr->FindValidationError(&s2err);
+        err << s2err.text();
+        throw py::value_error(err.str());
+    }
+
+    loop_ptr->Normalize();
+
+    return std::move(loop_ptr);
+}
+
+//
+// ---- Spherely Python Geography creation functions
+//
+
+PyObjectGeography point(double longitude, double latitude) {
+    return make_py_geography<s2geog::PointGeography>(make_s2point(longitude, latitude));
+}
+
+py::array_t<PyObjectGeography> points(const py::array_t<double> &coords) {
+    auto coords_data = coords.unchecked<2>();
+
+    if (coords_data.shape(1) != 2) {
+        throw std::runtime_error("coords array must be of shape (N, 2)");
+    }
+
+    auto npoints = coords_data.shape(0);
+    auto points = py::array_t<PyObjectGeography>(npoints);
+
+    py::buffer_info buf = points.request();
+    py::object *data = static_cast<py::object *>(buf.ptr);
+
+    for (py::ssize_t i = 0; i < npoints; i++) {
+        auto point_ptr = point(coords_data(i, 0), coords_data(i, 1));
+        data[i] = py::cast(std::move(point_ptr));
```

Comment:
For my education, why do we actually need the `py::cast` here? Because `point_ptr` is already a PyObjectGeography object, right?

Reply:
We don't need it, actually :-) (lines of code naively copied from old snippets).